File: python/links_client/api/recursive_links.py
"""RecursiveLinks - Recursive API wrapper for nested lists and dicts"""

from typing import List, Dict, Any, Optional, Union
from links_client.api.ilinks import ILinks
from links_client.utils.logger import get_logger

logger = get_logger(__name__)


class RecursiveLinks:
    """
    RecursiveLinks - Recursive wrapper for ILinks that supports nested structures

    Converts between Python nested lists/dicts and Links notation:
    - [[1, 2], [3, 4]] ↔ ((1 2) (3 4))
    - { "1": [1, { "2": [5, 6] }, 3, 4] } ↔ (1: 1 (2: 5 6) 3 4)
    """

    def __init__(self, db_path: Optional[str] = None):
        """
        Initialize RecursiveLinks

        Args:
            db_path: Path to database file
        """
        self.links = ILinks(db_path)
        self.id_counter = 1000000  # Start high to avoid conflicts with user IDs

    def get_links(self) -> ILinks:
        """
        Get the underlying ILinks instance

        Returns:
            ILinks instance
        """
        return self.links

    def _generate_temp_id(self) -> int:
        """
        Generate a unique temporary ID for internal use

        Returns:
            New unique ID
        """
        temp_id = self.id_counter
        self.id_counter += 1
        return temp_id

    def create_from_nested_list(self, nested_list: List[Any]) -> List[int]:
        """
        Create links from nested list structure
        [[1, 2], [3, 4]] represents two links: (1 2) and (3 4)

        Args:
            nested_list: Nested list structure

        Returns:
            List of created link IDs
        """
        try:
            link_ids = []

            for item in nested_list:
                if isinstance(item, list):
                    if len(item) >= 2:
                        # Create a link from [source, target]
                        source, target = item[0], item[1]

                        # Recursively handle nested lists
                        actual_source = (
                            self.create_from_nested_list([source])[0]
                            if isinstance(source, list)
                            else source
                        )
                        actual_target = (
                            self.create_from_nested_list([target])[0]
                            if isinstance(target, list)
                            else target
                        )

                        link_id = self.links.create([actual_source, actual_target])
                        link_ids.append(link_id)
                    else:
                        raise ValueError("List items must have at least 2 elements [source, target]")
                else:
                    # Single value - skip
                    logger.warning(f"Skipping non-list item in nested list: {item}")

            return link_ids
        except Exception as error:
            logger.error(f"Failed to create from nested list: {error}")
            raise

    def create_from_nested_dict(self, nested_dict: Dict[str, Any]) -> Dict[str, int]:
        """
        Create links from nested dict structure with references
        { "1": [1, { "2": [5, 6] }, 3, 4] } represents (1: 1 (2: 5 6) 3 4)

        Args:
            nested_dict: Nested dict structure with named links

        Returns:
            Map of reference names to created link IDs
        """
        try:
            reference_map = {}

            # Process each named link in the dict
            for ref_name, value in nested_dict.items():
                if isinstance(value, list):
                    # Create a sequence of links from the list
                    link_id = self._create_sequence_from_list(value, reference_map)
                    reference_map[ref_name] = link_id
                else:
                    logger.warning(f"Skipping non-list value in nested dict: {ref_name}={value}")

            return reference_map
        except Exception as error:
            logger.error(f"Failed to create from nested dict: {error}")
            raise

    def _create_sequence_from_list(
        self,
        lst: List[Any],
        reference_map: Dict[str, int]
    ) -> int:
        """
        Create a sequence of links from a list, handling nested dicts

        Args:
            lst: List with potential nested dicts
            reference_map: Map to store reference IDs

        Returns:
            ID of the created sequence link
        """
        if not lst:
            raise ValueError("Cannot create sequence from empty list")

        # For lists with exactly 2 elements, create a single link
        if len(lst) == 2 and not self._has_nested_dict(lst):
            source, target = lst[0], lst[1]
            return self.links.create([source, target])

        # For longer sequences or nested structures, create a chain of links
        current_id = None

        for item in lst:
            if isinstance(item, dict):
                # Nested dict with references
                nested_refs = self.create_from_nested_dict(item)
                # Use the first created link from nested dict
                first_ref = next(iter(nested_refs.values()))

                if current_id is None:
                    current_id = first_ref
                else:
                    # Create link connecting previous to this nested structure
                    current_id = self.links.create([current_id, first_ref])

                # Merge reference maps
                reference_map.update(nested_refs)
            elif isinstance(item, list):
                # Nested list
                nested_id = self._create_sequence_from_list(item, reference_map)

                if current_id is None:
                    current_id = nested_id
                else:
                    current_id = self.links.create([current_id, nested_id])
            else:
                # Simple value
                if current_id is None:
                    current_id = item
                else:
                    current_id = self.links.create([current_id, item])

        return current_id

    def _has_nested_dict(self, lst: List[Any]) -> bool:
        """
        Check if list contains nested dicts

        Args:
            lst: List to check

        Returns:
            True if contains nested dicts
        """
        return any(isinstance(item, dict) for item in lst)

    def read_as_nested_list(self, restriction: Optional[List[int]] = None) -> List[List[int]]:
        """
        Read links and convert to nested list structure

        Args:
            restriction: Filter for links

        Returns:
            Nested list structure
        """
        try:
            result = []
            visited = set()

            def handler(link):
                if link["id"] not in visited:
                    visited.add(link["id"])
                    result.append([link["source"], link["target"]])
                return self.links.constants.CONTINUE

            self.links.each(restriction, handler)

            return result
        except Exception as error:
            logger.error(f"Failed to read as nested list: {error}")
            raise

    def to_links_notation(self, nested_list: List[Any]) -> str:
        """
        Convert nested list to Links notation string
        [[1, 2], [3, 4]] -> "((1 2) (3 4))"

        Args:
            nested_list: Nested list structure

        Returns:
            Links notation string
        """
        def convert(item):
            if isinstance(item, list):
                inner = ' '.join(convert(el) for el in item)
                return f"({inner})"
            return str(item)

        inner = ' '.join(convert(item) for item in nested_list)
        return f"({inner})"

    def to_links_notation_with_refs(self, nested_dict: Dict[str, Any]) -> str:
        """
        Convert nested dict with references to Links notation string
        { "1": [1, { "2": [5, 6] }, 3, 4] } -> "(1: 1 (2: 5 6) 3 4)"

        Args:
            nested_dict: Nested dict structure

        Returns:
            Links notation string with references
        """
        def convert(item, ref_name=None):
            if isinstance(item, dict):
                # Nested dict with its own references
                parts = [convert(val, ref) for ref, val in item.items()]
                return ' '.join(parts)
            elif isinstance(item, list):
                inner = ' '.join(convert(el) for el in item)
                return f"({ref_name}: {inner})" if ref_name else f"({inner})"
            return str(item)

        parts = [convert(val, ref) for ref, val in nested_dict.items()]
        return f"({' '.join(parts)})"

    def parse_links_notation(self, notation: str) -> List[Any]:
        """
        Parse Links notation string to nested list
        "((1 2) (3 4))" -> [[1, 2], [3, 4]]

        Args:
            notation: Links notation string

        Returns:
            Nested list structure
        """
        notation = notation.strip()

        # Remove outer parentheses
        if notation.startswith('(') and notation.endswith(')'):
            notation = notation[1:-1].strip()

        result = []
        current = ''
        depth = 0

        for char in notation:
            if char == '(':
                depth += 1
                current += char
            elif char == ')':
                depth -= 1
                current += char

                if depth == 0 and current.strip():
                    result.append(self.parse_links_notation(current.strip()))
                    current = ''
            elif char == ' ' and depth == 0:
                if current.strip():
                    # Parse number
                    try:
                        num = int(current.strip())
                        result.append(num)
                    except ValueError:
                        pass
                    current = ''
            else:
                current += char

        # Handle remaining content
        if current.strip():
            try:
                num = int(current.strip())
                result.append(num)
            except ValueError:
                pass

        return result

File: js/src/api/recursive-links.js
// RecursiveLinks.js - Recursive API wrapper for nested arrays and objects
import ILinks from './ilinks.js';
import logger from '../utils/logger.js';

/**
 * RecursiveLinks - Recursive wrapper for ILinks that supports nested structures
 *
 * Converts between JavaScript nested arrays/objects and Links notation:
 * - [[1, 2], [3, 4]] ↔ ((1 2) (3 4))
 * - { "1": [1, { "2": [5, 6] }, 3, 4] } ↔ (1: 1 (2: 5 6) 3 4)
 */
class RecursiveLinks {
  constructor(dbPath = null) {
    this.links = new ILinks(dbPath);
    this.idCounter = 1000000; // Start high to avoid conflicts with user IDs
  }

  /**
   * Get the underlying ILinks instance
   * @returns {ILinks} - ILinks instance
   */
  getLinks() {
    return this.links;
  }

  /**
   * Generate a unique temporary ID for internal use
   * @private
   * @returns {number} - New unique ID
   */
  _generateTempId() {
    return this.idCounter++;
  }

  /**
   * Create links from nested array structure
   * [[1, 2], [3, 4]] represents two links: (1 2) and (3 4)
   *
   * @param {Array} nestedArray - Nested array structure
   * @returns {Promise<Array<number>>} - Array of created link IDs
   */
  async createFromNestedArray(nestedArray) {
    try {
      const linkIds = [];

      for (const item of nestedArray) {
        if (Array.isArray(item)) {
          if (item.length >= 2) {
            // Create a link from [source, target]
            const [source, target] = item;

            // Recursively handle nested arrays
            const actualSource = Array.isArray(source)
              ? (await this.createFromNestedArray([source]))[0]
              : source;
            const actualTarget = Array.isArray(target)
              ? (await this.createFromNestedArray([target]))[0]
              : target;

            const linkId = await this.links.create([actualSource, actualTarget]);
            linkIds.push(linkId);
          } else {
            throw new Error('Array items must have at least 2 elements [source, target]');
          }
        } else {
          // Single value - treat as a link to itself or skip
          logger.warn({ item }, 'Skipping non-array item in nested array');
        }
      }

      return linkIds;
    } catch (error) {
      logger.error({ error: error.message }, 'Failed to create from nested array');
      throw error;
    }
  }

  /**
   * Create links from nested object structure with references
   * { "1": [1, { "2": [5, 6] }, 3, 4] } represents (1: 1 (2: 5 6) 3 4)
   *
   * @param {object} nestedObject - Nested object structure with named links
   * @returns {Promise<object>} - Map of reference names to created link IDs
   */
  async createFromNestedObject(nestedObject) {
    try {
      const referenceMap = {};

      // Process each named link in the object
      for (const [refName, value] of Object.entries(nestedObject)) {
        if (Array.isArray(value)) {
          // Create a sequence of links from the array
          const linkId = await this._createSequenceFromArray(value, referenceMap);
          referenceMap[refName] = linkId;
        } else {
          logger.warn({ refName, value }, 'Skipping non-array value in nested object');
        }
      }

      return referenceMap;
    } catch (error) {
      logger.error({ error: error.message }, 'Failed to create from nested object');
      throw error;
    }
  }

  /**
   * Create a sequence of links from an array, handling nested objects
   * @private
   * @param {Array} arr - Array with potential nested objects
   * @param {object} referenceMap - Map to store reference IDs
   * @returns {Promise<number>} - ID of the created sequence link
   */
  async _createSequenceFromArray(arr, referenceMap) {
    if (arr.length === 0) {
      throw new Error('Cannot create sequence from empty array');
    }

    // For arrays with exactly 2 elements, create a single link
    if (arr.length === 2 && !this._hasNestedObject(arr)) {
      const [source, target] = arr;
      return await this.links.create([source, target]);
    }

    // For longer sequences or nested structures, create a chain of links
    // First element is the head
    let currentId = null;

    for (let i = 0; i < arr.length; i++) {
      const item = arr[i];

      if (typeof item === 'object' && !Array.isArray(item)) {
        // Nested object with references
        const nestedRefs = await this.createFromNestedObject(item);
        // Use the first created link from nested object
        const firstRef = Object.values(nestedRefs)[0];

        if (currentId === null) {
          currentId = firstRef;
        } else {
          // Create link connecting previous to this nested structure
          currentId = await this.links.create([currentId, firstRef]);
        }

        // Merge reference maps
        Object.assign(referenceMap, nestedRefs);
      } else if (Array.isArray(item)) {
        // Nested array
        const nestedId = await this._createSequenceFromArray(item, referenceMap);

        if (currentId === null) {
          currentId = nestedId;
        } else {
          currentId = await this.links.create([currentId, nestedId]);
        }
      } else {
        // Simple value
        if (currentId === null) {
          currentId = item;
        } else {
          currentId = await this.links.create([currentId, item]);
        }
      }
    }

    return currentId;
  }

  /**
   * Check if array contains nested objects
   * @private
   * @param {Array} arr - Array to check
   * @returns {boolean} - True if contains nested objects
   */
  _hasNestedObject(arr) {
    return arr.some(item => typeof item === 'object' && !Array.isArray(item));
  }

  /**
   * Read links and convert to nested array structure
   * @param {Array<number>|null} restriction - Filter for links
   * @returns {Promise<Array>} - Nested array structure
   */
  async readAsNestedArray(restriction = null) {
    try {
      const result = [];
      const visited = new Set();

      await this.links.each(restriction, async (link) => {
        if (!visited.has(link.id)) {
          visited.add(link.id);
          result.push([link.source, link.target]);
        }
        return this.links.constants.Continue;
      });

      return result;
    } catch (error) {
      logger.error({ error: error.message }, 'Failed to read as nested array');
      throw error;
    }
  }

  /**
   * Convert nested array to Links notation string
   * [[1, 2], [3, 4]] -> "((1 2) (3 4))"
   *
   * @param {Array} nestedArray - Nested array structure
   * @returns {string} - Links notation string
   */
  toLinksNotation(nestedArray) {
    const convert = (item) => {
      if (Array.isArray(item)) {
        const inner = item.map(convert).join(' ');
        return `(${inner})`;
      }
      return String(item);
    };

    return `(${nestedArray.map(convert).join(' ')})`;
  }

  /**
   * Convert nested object with references to Links notation string
   * { "1": [1, { "2": [5, 6] }, 3, 4] } -> "(1: 1 (2: 5 6) 3 4)"
   *
   * @param {object} nestedObject - Nested object structure
   * @returns {string} - Links notation string with references
   */
  toLinksNotationWithRefs(nestedObject) {
    const convert = (item, refName = null) => {
      if (typeof item === 'object' && !Array.isArray(item)) {
        // Nested object with its own references
        return Object.entries(item)
          .map(([ref, val]) => convert(val, ref))
          .join(' ');
      } else if (Array.isArray(item)) {
        const inner = item.map(el => convert(el)).join(' ');
        return refName ? `(${refName}: ${inner})` : `(${inner})`;
      }
      return String(item);
    };

    const parts = Object.entries(nestedObject).map(([ref, val]) => convert(val, ref));
    return `(${parts.join(' ')})`;
  }

  /**
   * Parse Links notation string to nested array
   * "((1 2) (3 4))" -> [[1, 2], [3, 4]]
   *
   * @param {string} notation - Links notation string
   * @returns {Array} - Nested array structure
   */
  parseLinksNotation(notation) {
    notation = notation.trim();

    // Remove outer parentheses
    if (notation.startsWith('(') && notation.endsWith(')')) {
      notation = notation.slice(1, -1).trim();
    }

    const result = [];
    let current = '';
    let depth = 0;

    for (let i = 0; i < notation.length; i++) {
      const char = notation[i];

      if (char === '(') {
        depth++;
        current += char;
      } else if (char === ')') {
        depth--;
        current += char;

        if (depth === 0 && current.trim()) {
          result.push(this.parseLinksNotation(current.trim()));
          current = '';
        }
      } else if (char === ' ' && depth === 0) {
        if (current.trim()) {
          // Parse number
          const num = parseInt(current.trim());
          if (!isNaN(num)) {
            result.push(num);
          }
          current = '';
        }
      } else {
        current += char;
      }
    }

    // Handle remaining content
    if (current.trim()) {
      const num = parseInt(current.trim());
      if (!isNaN(num)) {
        result.push(num);
      }
    }

    return result;
  }
}

export default RecursiveLinks;
